### DISCUSSION TITLE
Given Word and Synset how to get the Sense

### DISCUSSION THREAD

**Discussion by vsraptor** [goodmami/wn#135] [Q&A] [answered]
i'm looping trought word synsets ...afaik there is only one sense per pair word ==sense==> synset !  right ?

how can i get it ?

```
	for wsyn in wn.synsets(word):
		key = hash(wsyn.sense(word))  <=== singular
		senses.append( sense_map[key] )
```

As a side question is this the fastest way to check for the position of the sense ... i dont want to use dict if possible (wondering if hash() is necessary)

`sense_map = { hash(e):i for i,e in enumerate(wn.senses()) }`

**Comment by goodmami** [comment, marked as answer]
> there is only one sense per pair word ==sense==> synset ! right ?

A sense is an edge between a word and a synset showing that a word has the conceptual meaning captured by the synset (there is some documentation [here](https://wn.readthedocs.io/en/latest/guides/wordnet.html)). There may be many senses for a word, and many words for a synset. While it is technically possible for the same word and synset to have multiple sense links between them, such a situation would be redundant and erroneous, and that wordnet should be corrected.

> how can i get it ?

By "it" I think you mean getting the sense, if any, between a given `wn.Word` and `wn.Synset`. There is not currently a way to get the sense when you already have the word and synset objects, although I see how that can be useful. If by "word" you mean "wordform" and not `wn.Word` object, then you can do something like:

```python
for sense in wn.senses(word):
    word = sense.word()
    synset = sense.synset()
```

You could use these to populate a dictionary for later lookup, if you like:

```python
    lookup_dict[(word, synset)] = sense
```

> As a side question is this the fastest way to check for the position of the sense ... i dont want to use dict if possible (wondering if hash() is necessary)

I'm not sure what you mean by "position". If you mean a unique key for use in a dictionary, the `wn.Word`, `wn.Sense`, and `wn.Synset` objects are all hashable (you don't need to call `hash()` explicitly), so you can just do:

```python
d = {}
for sense in wn.senses(word):
    d[sense] = (sense.word(), sense.synset())
```

But if you know you're only working within a single lexicon, you can use the identifiers, which is a bit faster if you're doing this frequently:

```python
d = {}
for sense in wn.senses(word):
    d[sense.id] = (sense.word(), sense.synset())
```

While it's unlikely, a synset, sense, or word identifier can be reused in different lexicons, but not within a single lexicon. When you use `wn.senses()` (`wn.words()`, `wn.synsets()`, etc.), you are querying any lexicons that are installed. To restrict the query to a single lexicon, you can specify it in the query:

```python
wn.senses(word, lexicon='pwn:3.0')  # or whatever lexicon you want
```

Or, even better, create a `wn.Wordnet` object and use that:

```python
pwn = wn.Wordnet(lexicon='pwn:3.0')
pwn.senses(word)
```

**Reply by vsraptor** [reply]
> By "it" I think you mean getting the sense, if any, between a given `wn.Word` and `wn.Synset`. There is not currently a way to get the sense when you already have the word and synset objects, although I see how that can be useful. If by "word" you mean "wordform" and not `wn.Word` object, then you can do something like:
> 
> ```python
> for sense in wn.senses(word):
>     word = sense.word()
>     synset = sense.synset()
> ```
> 
> You could use these to populate a dictionary for later lookup, if you like:
> 
> ```python
>     lookup_dict[(word, synset)] = sense
> ```

I meant wordform not word obj, but i suppose u can get back word-obj from word form.

So the only way is to enumerate ;(... any SQL trick ? ;)

**Reply by vsraptor** [reply]
> I'm not sure what you mean by "position". If you mean a unique key for use in a dictionary, the `wn.Word`, `wn.Sense`, and `wn.Synset` objects are all hashable (you don't need to call `hash()` explicitly), so you can just do:

sorry, i meant what i just did simply sense <=> unique-integer map.
i generate it once and then use it..

i'm ok thanks

**Reply by vsraptor** [reply]
>  (there is some documentation [here](https://wn.readthedocs.io/en/latest/guides/wordnet.html)).

i read it many times .. it is good but is hard to wrap your head around it..

I read the "edge" idea somewhere else in the docs. That made it much more clear ..

It is a bit confusing because in most of philosophy/psychology thesis's you have just word <==> sense/concept mapping ... wordnet add the Synsets , the edge-idea made it clear.

Do you happen to know what is the underlining explanation of why ? it have to explain some fundamental structure of language or the brain ... just wondering aloud

**Reply by goodmami** [reply]
> So the only way is to enumerate ;(... any SQL trick ? ;)

Even SQL would enumerate to find the relevant data. It would just be faster. If you want O(1) lookup then the answer is to create a dict mapping what you need. Creating this would require enumerating all the senses once.

> i read it many times .. it is good but is hard to wrap your head around it..

That part of the documentation (along with another section or two) is incomplete and could use some editing, so I understand if it's a bit hard to follow. I welcome suggestions to improve them.

> Do you happen to know what is the underlining explanation of why ? it have to explain some fundamental structure of language or the brain ... just wondering aloud

I don't think wordnets ever had claims to model brain structures. I think it's just the way things are encoded in [LMF](https://globalwordnet.github.io/schemas/) wordnets. The older [WNDB](https://wordnet.princeton.edu/documentation/wndb5wn) format does not use senses, at least, not as distinct entities in the data. Having explicit senses makes it easier to encode things like sense relations, etc.
